collectors/collector.go

```diff
@@ -0,0 +1,165 @@
+package collectors
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"time"
+)
+
+// Collector is the interface to abstract data collection from different sources
+type Collector interface {
+	// Name returns the Name of this collector
+	Name() string
+	// Optional returns true if Collector is not mandatory and fails should be ignored
+	Optional() bool
+	// Collect returns collected data
+	Collect(ctx context.Context) (io.ReadCloser, error)
+}
+
+type NameOptional struct {
+	Name     string
+	Optional bool
+}
+
+type CmdCollector struct {
+	NameOptional
+	Cmd []string
+}
+
+func (c CmdCollector) Name() string {
+	return c.NameOptional.Name
+}
+
+func (c CmdCollector) Optional() bool {
+	return c.NameOptional.Optional
+}
+
+func (c CmdCollector) Collect(ctx context.Context) (io.ReadCloser, error) {
+	cmd := exec.CommandContext(ctx, c.Cmd[0], c.Cmd[1:]...)
+	output, err := cmd.CombinedOutput()
+	return ioutil.NopCloser(bytes.NewReader(output)), err
+}
+
+// TODO(janisz): Make use of this code instead of calling dcos-diagnostics for units data
```

Comment:
I'm wondering: Shall we compose a list of all the TODOs you're adding with this PR and create JIRA tasks for the most important ones? Otherwise I suspect these will stay here forever.

Reply:
Done